README.md
XML2R
=======

EasieR XML data collection. For an introduction to the package, see [here](http://cpsievert.github.io/XML2R/)

To install the [CRAN](http://cran.r-project.org/web/packages/XML2R/index.html) version use: `install.packages("XML2R"); library(XML2R)`

To install the github version use: `library(devtools); install_github("XML2R", "cpsievert"); library(XML2R)`

TO DO:

(1) Fix naming of observations when no children exist   
(2) Add some tests using testthat   

NEWS
    CHANGES IN XML2R VERSION 0.0.7

MAJOR FIX

  - Local XML files are once again supported. See the local argument in XML2Obs/urlsToDocs.

MINOR CHANGES

  - The async (asynchronous download) option was deprecated as httr:GET is now used instead of RCurl::getURL for more flexible http/https requests


    CHANGES IN XML2R VERSION 0.0.6

MINOR FIX

  - In version 0.5, the url column does not contain the file name (instead, it contains integers). This has consequences for pitchRx, but is now fixed.

    CHANGES IN XML2R VERSION 0.0.5

MAJOR CHANGES/IMPROVEMENTS

  - A dependency to RCurl was added so asynchronous downloads can be done via getURL. This dependency also has the benefit of more robust HTTP downloads.

    CHANGES IN XML2R VERSION 0.0.4

MAJOR CHANGES/IMPROVEMENTS

  - The collapse() function was changed to collapse_obs() to avoid namespace clashing with dplyr's collapse().

  - The child argument was removed from add_key(). The recycle argument was added to add_key() which allows one to use an existing value in the parent node as a key to connect the parent observation to its descendents (thanks Carlos Scheidegger) 

  - Better garbage collection

    CHANGES IN XML2R VERSION 0.0.3

IMPROVEMENTS

  - url.map option was added to XML2Obs.

    CHANGES IN XML2R VERSION 0.0.2

MAJOR CHANGES

  - Function rename() was changed to re_name() to avoid NAMESPACE issue conflicts with plyr.

MINOR FIXES

  - quiet option added to several functions

    CHANGES IN XML2R VERSION 0.0.1

NEW FEATURES

  - First version of XML2R. See the package [tutorial](http://cpsievert.github.io/XML2R/)

  - This package was born out of an effort to abstract the functionality of [pitchRx::urlsToDataFrame](https://github.com/cpsievert/pitchRx/blob/master/R/urlsToDataFrame.R) so that other projects can be built on top of this framework.

MISC

  - in this NEWS file, #n means the issue number on GitHub, e.g. #1 is
  https://github.com/cpsievert/XML2R/issues/1


R/utils.R
#' Parse XML Files into XML Documents
#' 
#' Essentially a recursive call to \link{xmlParse}.
#' 
#' @param urls character vector. Either urls that point to an XML file online or a local XML file name.
#' @param local logical. Should urls be treated as paths to local files?
#' @param quiet logical. Print file name currently being parsed?
#' @param ... arguments passed along to \link{httr::GET}
#' @importFrom plyr try_default
#' @importFrom XML xmlParse
#' @importFrom httr url_ok
#' @importFrom httr GET
#' @importFrom httr content
#' @export
            
urlsToDocs <- function(urls, local = FALSE, quiet = FALSE, ...) {
  #keep only urls that exist
  if (!local) {
    urls <- urls[vapply(urls, url_ok, logical(1), USE.NAMES=FALSE)]
    text <- lapply(urls, function(x) content(GET(x, ...), as = "text"))
  } else {
    text <- urls
  }
  if (length(text) == 0) {
    warning("No content found. Please double check your urls.")
    return(text)
  }
  docs <- NULL
  for (i in seq_along(text)) {
    if (!quiet) cat(urls[i], "\n")
    doc <- try_default(xmlParse(text[i], asText = !local), NULL, quiet = TRUE)
    if (!is.null(doc)) {
      attr(doc, "XMLsource") <- urls[i]
      docs <- c(docs, doc) #Keep non-empty documents
    }
  }
  return(docs)
}

#' Parse XML Documents into XML Nodes
#' 
#' Essentially a recursive call to \link{getNodeSet}.
#' 
#' @param docs XML documents
#' @param xpath xpath expression
#' @importFrom XML getNodeSet
#' @export

docsToNodes <- function(docs, xpath) {
  #I should really figure which class I want...
  rapply(docs, function(x) getNodeSet(x, path=xpath), 
         classes=c('XMLInternalDocument', 'XMLAbstractDocument'), how="replace")
}


#' Coerce XML Nodes into a list with both attributes and values
#' 
#' Essentially a recursive call to \link{xmlToList}.
#' 
#' @param nodes A collection of XML nodes. Should be the output from \link{docsToNodes}.
#' @importFrom XML xmlToList
#' @return A nested list with a structure that resembles the XML structure
#' @export
#' 

nodesToList <- function(nodes){
  #I should really figure which class I want...
  rapply(nodes, function(x) xmlToList(x),
    classes=c("XMLInternalElementNode", "XMLInternalNode", "XMLAbstractNode"), how="replace")
}

#' Flatten nested list into a list of observations
#' 
#' This function flattens the nested list into a list of "observations" (that is, a list of matrices with one row).
#' The names of the list that is returned reflects the XML ancestory of each observation.
#' 
#' @param l list. Should be the output from \link{nodesToList}. 
#' @param urls character vector the same length as \code{l}. Each element should map element of \code{l} to an XML file.
#' @param append.value logical. Should the XML value be appended to the observation?
#' @param as.equiv logical. Should observations from two different files (but the same ancestory) have the same name returned?
#' @param url.map logical. If TRUE, the 'url_key' column will contain a condensed url identifier (for each observation)
#' and full urls will be stored in the "url_map" element. If FALSE, the full urls are included (for each observation) 
#' as a 'url' column and no "url_map" is included.
#' @return A list where each element reflects one "observation".
#' @export

#adapted from knowledege gained from here:
#http://stackoverflow.com/questions/8139677/how-to-flatten-a-list-to-a-list-without-coercion?
##http://stackoverflow.com/questions/18862601/extract-name-hierarchy-for-each-leaf-of-a-nested-list

listsToObs <- function(l, urls, append.value=TRUE, as.equiv=TRUE, url.map=TRUE) {
  #add a prefix to the names of the list to help content to a url
  url.count <- paste0("url", seq_len(length(urls)))
  url_map <- cbind(url_key=url.count, url=urls)
  names(l) <- url.count
  #Assuming the names of the list elements holding XML values (and only values) will be NULL, we can distinguish between values/attributes
  #By also assuming that values appear immediately before their respective attributes (which appears to be the way xmlToList works),
  #we append the XML value to a row of attributes (given that their from the same node)
  name.len <- rapply(l, function(x) length(names(x)))  
  list.len <- length(name.len)
  nms <- names(name.len)
  #if (is.null(names(l[[1]]))) nms <- rep(names(l), sapply(l, length)) #overwrite names if no names (or no children) exist
  # turn '..attr' into '.attr' -- note it can show up multiple times in a name
  temp <- gsub('.attrs', 'attrs', fixed=TRUE, nms)
  idx <- gsub('.', '//', fixed=TRUE, temp)
  #select the url prefix and replace with nothing (this will give observations from different files and the same ancestory the same name)
  if (as.equiv) {
    node.sets <- sub("url([0-9]+)//", "", idx)
  } else {
    node.sets <- idx
  }
  urlz <- sub("//.*$", "", idx)
  suffix <- sub(".*//", "", node.sets)
  #tracks which XML values should be appended to the sequential row
  indicies <- which(name.len == 0 & suffix %in% "text") 
  holder <- vector('list', list.len) #placeholder for the flattened list hierarchy
  i <- 0L
  #fill up placeholder with relevant info
  rapply(l, function(x) { 
              i <<- i+1L 
              holder[[i]] <<- matrix(x, nrow=1)
              nmz <- names(x)
              if (is.null(nmz)) nmz <- "XML_value"
              colnames(holder[[i]]) <<- nmz
            })
  #Append XML_value column to the appropriate attributes
  #Note that this assumes the value always appears before the attributes in the list order
  if (append.value && length(indicies) > 0) {
    values <- holder[indicies]
    add.values <- holder[indicies+1]
    holder[indicies+1] <- mapply(function(x, y){ cbind(x, y) }, add.values, values, SIMPLIFY=FALSE)
    holder[indicies] <- NULL
    #remove the XML value elements (so they aren't duplicated)
    urlz <- urlz[-indicies]
    node.sets <- node.sets[-indicies]
  }
  if (url.map) {
    holder <- mapply(function(x, y) cbind(x, url_key=y), holder, urlz, SIMPLIFY=FALSE)
    holder[[length(holder) + 1]] <- url_map #append url_map to end of list
    node.sets <- c(node.sets, "url_map")
  } else {
    tab <- table(urlz)
    reps <- tab[match(url.count, names(tab))]
    reps[is.na(reps)] <- 0 #error handling for urls without any observations
    holder <- mapply(function(x, y) cbind(x, url=y), holder, rep(urls, reps), SIMPLIFY=FALSE)
  }
  names(holder) <- sub("//attrs", "", node.sets)
  return(holder)
}


#' Rename rows of a list
#' 
#' Sometimes, certain nodes in an XML ancestory may want to be neglected
#' before any keys are created (see \link{add_key}) or observations are aggregated (see \link{collapse}).
#' This function takes a list of "observations" (that is, a list of matrices with one row) and 
#' alters the names of that list. Note that any information lost from changing names is saved 
#' in a new column whose name is specified by \code{diff.name}.
#' 
#' @param obs list. Should be the output from \link{XML2Obs} (or \link{listsToObs}). 
#' @param namez must be equivalent to \code{names(obs)}. Intended use is to avoid unneccessarily repeating that operation.
#' @param equiv character vector with the appropriate (unique) names that should be regarded "equivalent".
#' @param diff.name character string used for naming the variable that is appended to any observations whose name was overwritten. 
#' The value for this variable is the difference in from the original name and the overwritten name.
#' @param rename.as character string to override naming of observations that are renamed.
#' @param quiet logical. Include message about how observations are being renamed?
#' @return A list of "observations". 
#' @export

re_name <- function(obs, namez, equiv, diff.name="diff_name", rename.as, quiet=FALSE){
  if (missing(equiv)) {
    warning("Must include equiv argument!")
    return(obs)
  }
  if (missing(namez)) {
    nms <- names(obs)
  } else {
    nms <- namez
  }
  if (is.null(nms)) {
    warning("The observations don't not have any names!")
    return(obs)
  }
  if (all(!equiv %in% unique(nms))) {
    warning("None of the equiv elements match the names of the observations.")
    return(obs)
  }
  baseline <- strsplit(equiv, "//")
  n <- length(equiv)
  keeps <- baseline[[1]]
  names(baseline) <- equiv
  for (i in seq_len(n-1)) { #get all the common nodes
    idx <- keeps %in% baseline[[i+1]]
    keeps <- keeps[idx]
  }
  #construct the 'label' to replace all the names that match equiv
  if (length(keeps) == 0) {
    warning("None of the XML node names seem to match. Are you sure you want to regard these names equivalent?")
    label <- "equivalent"
  } else {
    label <- paste0(keeps, collapse="//") 
  }
  if (!missing(rename.as)) label <- rename.as
  if (!quiet) message(paste0("Renaming all list elements named: \n", paste(equiv, collapse="  OR  "), "\nwith\n", label))
  diffs <- lapply(baseline, function(x) paste(x[!x %in% keeps], collapse="//")) #keeps the nodes that will be 'overwritten'
  idx <- nms %in% equiv
  #idx <- grepl(paste(equiv, collapse="||"), nms)  #grep here instead?
  #overwrite the names
  names(obs)[idx] <- label
  #get the information that was "lost" when overwritting names and append a new column accordingly
  klass <- as.character(diffs[match(nms[idx], names(diffs))]) 
  obs[idx] <- mapply(function(x, y) cbind(x, `colnames<-`(cbind(y), diff.name)), obs[idx], klass, SIMPLIFY=FALSE)
  return(obs)
}

#' Add a key to connect parents to descendants
#' 
#' This function creates a mapping from parent observations to it's descendants (which useful for merging/joining tables).
#' Either an existing value in the parent observation can be \code{recycle}d to it's descendants or a new column 
#' will be created (if \code{recycle} is missing).
#' 
#' @param obs list. Should be the output from \link{listsToObs}. 
#' @param parent character string. Should be present in the names of \code{obs}.
#' @param recycle character string that matches a variable name among \code{parent} observations.
#' @param key.name The desired column name of the newly generated key.
#' @param quiet logical. Include message about the keys being generated?
#' @return A list of observations.
#' @export

add_key <- function(obs, parent, recycle, key.name, quiet=FALSE){
  if (missing(parent)) {
    warning("You must provide the parent argument!")
    return(obs)
  }
  if (length(parent) > 1) warning("Please specify one parent at a time!")
  nms <- names(obs)
  if (is.null(nms)) {
    warning("The observations don't not have any names!")
    return(obs)
  }
  un <- unique(nms)
  if (!parent %in% un) {
    warning("The parent argument you provided does not match any observations.")
    return(obs)
  }
  fetus <- un[-which(un == parent)]
  children <- fetus[grep(paste0(parent, "//.*"), fetus)]
  if (length(children) == 0){
    warning(paste0("No children were found for the ", parent, " node."))
    return(obs)
  } else {
    if (!quiet) message(paste0("A key for the following children will be generated for the ", parent, " node:\n", paste0(children, collapse="\n")))
  }
  if (missing(recycle)) { #add an (outer) index to parent
    elders <- nms == parent
    outer_index <- seq_len(sum(elders))
    if (missing(key.name)) key.name <- "key_name"
    obs[elders] <- mapply(function(x, y) cbind(x, `colnames<-`(cbind(y), key.name)), obs[elders], outer_index, SIMPLIFY=FALSE) 
  } else { #use a specified variable for (outer) index
    elders <- nms == parent
    outer_index <- lapply(obs[elders], function(x) x[,recycle])
    if (missing(key.name)) key.name <- recycle
  }
  #now add the (inner) index for each descendent
  for (child in children) {
    kids <- nms == child
    kid.idx <- which(kids)
    relevant <- obs[elders | kids] #have to subset to relevant cases to get the proper indexing
    elder.idx <- which(names(relevant) == parent)
    timez <- diff(c(0, elder.idx))-1 #yields the number of children between each parent
    inner_index <- rep(outer_index, timez)
    obs[kid.idx] <- mapply(function(x, y) cbind(x, `colnames<-`(cbind(y), key.name)), obs[kid.idx], inner_index, SIMPLIFY=FALSE) 
  }
  return(obs)
}

#' Collapse a list of observations into a list of tables.
#' 
#' This function aggregates all observations with a similar name into a common table. Note that observations 
#' with a particular name don't need consistent variables (any missing information is filled with NAs).
#' 
#' @param obs list of observations.
#' @return Returns list with one element for each relevant XML node. Each element contains a matrix.
#' @importFrom plyr rbind.fill.matrix
#' @export

collapse_obs <- function(obs) {
  nms <- names(obs)
  #Exclude url_map from collapsing if it exists
  map <- grep("url_map", nms)
  if (length(map) != 0) {
    obs <- obs[-map]
    nms <- nms[-map] 
  }
  if (length(unique(nms)) == 1) {
    return(rbind.fill.matrix(obs))
  } else {
    return(tapply(obs, INDEX=nms, rbind.fill.matrix))
  }
}


R/urlsToDataFrame.R
#' Parse XML files into (a list of) matrices or data frame(s)
#' 
#' The main function of the XML2R package. Documentation to come...
#' 
#' @param urls character vector or list of urls that point to an XML file (or anything readable by \link{xmlParse}).
#' @param xpath XML XPath expression that is passed to \link{getNodeSet}. If missing, the entire root and all descendents are captured and returned (ie, tables = "/"). 
#' @param df logical. Should matrices be coerced into data frames?
#' @return Returns list with one element for each relevant XML node. Each element contains a matrix by default.
#' @seealso \link{urlsToDocs}, \link{docsToNodes}, \link{nodesToList}, \link{listsToMatrix}
#' @export
#' @examples
#' \dontrun{
#' pre <- "http://gd2.mlb.com/components/game/mlb/year_2013/month_06/day_14/"
#' post <- c("gid_2013_06_14_phimlb_colmlb_1/inning/inning_all.xml",
#'            "gid_2013_06_14_seamlb_oakmlb_1/inning/inning_all.xml")
#' urls <- paste0(pre, post)
#' #collects everything
#' dat <- XML2R(urls) 
#' names(dat)
#' dat[[1]] #game level
#' dat[[2]] #inning level
#' head(dat[[3]]) #action level
#' head(dat[[4]]) #atbat level
#' #collects everything at the atbat level and lower (ie, pitch, runner, etc)
#' dat.atbat <- XML2R(urls, xpath="//atbat") 
#' head(dat.atbat[[1]])
#' 
#' urls2 <- c("http://gd2.mlb.com/components/game/mlb/year_2013/mobile/346180.xml",
#'            "http://gd2.mlb.com/components/game/mlb/year_2013/mobile/346188.xml")
#' dat3 <- XML2R(urls2)
#' 
#' cens <- "http://www.census.gov/developers/data/sf1.xml"
#' census <- XML2R(cens)
#' }
#' 

XML2R <- function(urls, xpath, df=FALSE) {
  if (missing(xpath)) xpath <- "/"
  docs <- urlsToDocs(urls)
  valid.urls <- sapply(docs, function(x) attr(x, "XMLsource"))
  nodes <- docsToNodes(docs, xpath) #select the root
  rm(docs)
  l <- nodesToList(nodes)
  rm(nodes)
  #since listsToMatrix "collapses" unnamed lists, we now add url identifiers to the first level of elements
  url.count <- paste0("url", seq_len(length(valid.urls)))
  names(valid.urls) <- url.count
  names(l) <- url.count
  m <- listsToMatrix(l)
  names(m) <- sub("//attrs", "", names(m))
  rm(l)
  if (df) m <- lapply(m, function(x) data.frame(x, stringsAsFactors=FALSE))
  m[["url_map"]] <- valid.urls
  return(m)
}

#' Collapse a list of matrices (or data frames) into a smaller list of larger matrices or data frames
#' 
#' @param m list of matrices or data frames
#' @param condition A partial match of names(m). 
#' @return Returns list with one element for each relevant XML node. Each element contains a matrix by default.
#' @importFrom plyr rbind.fill.matrix
#' @export

collapse <- function(m, condition="pitch") {
  #collapse list of matrices according to some type of syntax
  nms <- names(m)
  nmz <- nms[grep(condition, nms)]
  browser()
  nm.split <- strsplit(nmz, split="\\$")
  idx <- sapply(nm.split, function(x) x[grep(condition, x)])
  return(tapply(m, INDEX=idx, rbind.fill.matrix))
  #cond <- paste0(".*\\$$", condition)
  #gsub(".*\\$$")
  #i <- 0L
  #lapply(m, function(x) { i <<- i+1L; x$source <- names(m)[[i]] })
}


R/XML2R.R
#' Parse XML files into (a list of) matrices or data frame(s)
#' 
#' This function is an experimental wrapper around \link{XML2Obs}. One should only use this function over \link{XML2Obs} if 
#' keys already exist in the XML data and ancestory doesn't need to be altered.
#' 
#' @param urls character vector or list of urls that point to an XML file (or anything readable by \link{xmlParse}).
#' @param xpath XML XPath expression that is passed to \link{getNodeSet}. If missing, the entire root and all descendents are captured and returned (ie, tables = "/"). 
#' @param df logical. Should matrices be coerced into data frames?
#' @return Returns list with one element for each relevant XML node. Each element contains a matrix by default.
#' @seealso \link{urlsToDocs}, \link{docsToNodes}, \link{nodesToList}, \link{listsToObs}
#' @export
#' @examples
#' \dontrun{
#' urls2 <- c("http://gd2.mlb.com/components/game/mlb/year_2013/mobile/346180.xml",
#'            "http://gd2.mlb.com/components/game/mlb/year_2013/mobile/346188.xml")
#' dat3 <- XML2R(urls2)
#' 
#' cens <- "http://www.census.gov/developers/data/sf1.xml"
#' census <- XML2R(cens)
#' }
#' 

XML2R <- function(urls, xpath, df=FALSE) {
  obs <- XML2Obs(urls, xpath)
  #add an option to try to automatically generate keys?
  collapse_obs(obs)
}

#' Parse XML files into a list of "observations"
#' 
#' This function takes a collection of urls that point to XML files and coerces the relevant info into a list of observations.
#' An "observation" is defined as a matrix with one row. An observation can also be thought of as a single instance of 
#' XML attributes (and value) for a particular level in the XML hierarchy. The names of the list reflect the XML node 
#' ancestory for which each observation was extracted from.
#' 
#' It's worth noting that a "url_key" column is appended to each observation to help track the origin of each observation.
#' The value of the "url_key" column is not the actual file name, but a simplified identifier to avoid unnecessarily repeating 
#' long file names for each observation. For this reason, an addition element (named "url_map") is added to the list of observations
#' in case the actual file named want to be used.
#' 
#' @param urls character vector. Either urls that point to an XML file online or a local XML file name.
#' @param xpath XML XPath expression that is passed to \link{getNodeSet}. If missing, the entire root and all descendents are captured and returned (ie, tables = "/"). 
#' @param append.value logical. Should the XML value be appended for relevant observations?
#' @param as.equiv logical. Should observations from two different files (but the same ancestory) have the same name returned?
#' @param url.map logical. If TRUE, the 'url_key' column will contain a condensed url identifier (for each observation)
#' and full urls will be stored in the "url_map" element. If FALSE, the full urls are included (for each observation) 
#' as a 'url' column and no "url_map" is included.
#' @param local logical. Should urls be treated as paths to local files?
#' @param quiet logical. Print file name currently being parsed?
#' @param ... arguments passed along to \link{httr::GET}
#' @seealso \link{urlsToDocs}, \link{docsToNodes}, \link{nodesToList}, \link{listsToObs}
#' @return A list of "observations" and (possibly) the "url_map" element. 
#' @export
#' @examples
#' 
#' \dontrun{
#' urls <- c("http://gd2.mlb.com/components/game/mlb/year_2013/mobile/346180.xml",
#'            "http://gd2.mlb.com/components/game/mlb/year_2013/mobile/346188.xml")
#' obs <- XML2Obs(urls)
#' table(names(obs))
#' 
#' # parses local files as well
#' players <- system.file("extdata", "players.xml", package = "XML2R")
#' obs2 <- XML2Obs(players, local = TRUE)
#' table(names(obs2))
#' }

XML2Obs <- function(urls, xpath, append.value = TRUE, as.equiv = TRUE, url.map = FALSE, 
                    local = FALSE, quiet = FALSE, ...) {
  if (missing(xpath)) xpath <- "/"  #select the root
  docs <- urlsToDocs(urls, local, quiet, ...)
  valid.urls <- sapply(docs, function(x) attr(x, "XMLsource"))
  nodes <- docsToNodes(docs, xpath) 
  rm(docs)
  gc()
  l <- nodesToList(nodes)
  rm(nodes)
  gc()
  obs <- listsToObs(l, urls = valid.urls, append.value = append.value, 
                    as.equiv = as.equiv, url.map = url.map)
  return(obs)
}
